docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,1027 @@
+
+It may be evident to readers that there is some inherent but fuzzy connection
+between the idea of writing in the future and closing time in the future. It is
+interesting to point out that in systems like Spanner and YB which assign
+monotonically increasing timestamps to all writes within a replication group and
+therefore "instantaneously" close out time, these two operations are one and the
+same.
+
+CockroachDB's transaction and replication model are not quite set up to support
+this. Transactions in CockroachDB are optimized for committing at their original
+read timestamp, so they don't like when their writes get bumped to higher
+timestamps. This allows for optimistic, lockless reads. It also avoids a layer
+of write amplification, wherein provisional values must be moved to a higher
+timestamp during intent resolution. This write amplification is concerning
+enough that systems like Google's Percolator introduce a specialized storage
+component (see the "write metadata" column) to perform the translation without
+needing to rewrite the data [6].
+
+Additionally, replication in CockroachDB is below the level of evaluation (see
+[proposer evaluated kv](20160420_proposer_evaluated_kv.md)), so the timestamp of
+a replicated value need to be determined before replication-level sequencing.
+This allows for evaluation-level parallelism and less need for determinism
+during evaluation. This evaluation-level parallelism helps avoid the problem of
+the "parallelism gap" seen between compute and/or storage intensive processing
+before replication and after replication [7].
+
+For these reasons, the closed timestamp of a Range lags the timestamps of writes
+in a Range and the closed timestamp is more of a Range-level concern and less of
+a Request-level concern in CockroachDB (i.e. a single request cannot dictate the
+closed timestamp for a Range). This is why the closed timestamp dictates the
+time at which requests perform writes (as we'll see in the next section) instead
+of the other way around.
+
+### Becoming a Non-Blocking Transaction
+
+As stated earlier, this proposal does not expose "non-blocking transactions" to
+users directly. Instead, it exposes "consistent read Ranges" to users. It was
+also mentioned earlier that a "consistent read Range" will have a closed
+timestamp tracker that is leading "present time" by some duration. As it turns
+out, configuring this closed timestamp tracker to lead "present time" by
+`non_blocking_duration` is enough to implement "non-blocking transactions"
+without any other server-side changes.
+
+Any standard transaction that writes to a "consistent read Range" will naturally
+get pushed into the future. If it had performed writes in the past, these will
+eventually get moved up to the new commit timestamp. If it had performed reads
+in the past, these will eventually get refreshed up to the new (synthetic)
+commit timestamp before committing. The new Commit Wait logic in the
```

Comment:
wouldn't this cause any writes in the near future to the keys read by these refreshed reads (even for keys not part of a consistent read range) to block (similar to my question above about `txn2`)?
-- Edit: I think you covered this in "Non-blocking and Standard Transaction Interactions", but might be worth adding a pointer here.